pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -103,206 +141,300 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
+	retry := false
+	defer func() {
+		// Requeue or reset our retry count
+		if retry {
+			e.retry(e.syncSecretQueue, key)
+		} else {
+			e.retryCounts.Remove(key)
+		}
+		e.syncSecretQueue.Done(key)
+	}()
+
+	secretInfo, err := parseSecretQueueKey(key)
 	if err != nil {
 		glog.Error(err)
 		return
 	}
-	if serviceAccount == nil {
-		return
-	}
 
-	if err := client.RetryOnConflict(RemoveTokenBackoff, func() error {
-		return e.removeSecretReferenceIfNeeded(serviceAccount, secret.Name)
-	}); err != nil {
-		utilruntime.HandleError(err)
+	secret, err := e.getSecret(secretInfo.namespace, secretInfo.name, secretInfo.uid, false)
+	switch {
+	case err != nil:
+		glog.Error(err)
+		retry = true
+	case secret == nil:
+		// If the service account exists
+		if sa, saErr := e.getServiceAccount(secretInfo.namespace, secretInfo.saName, secretInfo.saUID, false); saErr == nil && sa != nil {
+			// secret no longer exists, so delete references to this secret from the service account
+			if err := client.RetryOnConflict(RemoveTokenBackoff, func() error {
+				return e.removeSecretReference(secretInfo.namespace, secretInfo.saName, secretInfo.saUID, secretInfo.name)
+			}); err != nil {
+				glog.Error(err)
+			}
+		}
+	default:
+		// Ensure service account exists
+		sa, saErr := e.getServiceAccount(secretInfo.namespace, secretInfo.saName, secretInfo.saUID, true)
+		switch {
+		case saErr != nil:
+			glog.Error(saErr)
+			retry = true
+		case sa == nil:
+			// Delete token
+			glog.V(4).Infof("syncSecret(%s/%s), service account does not exist, deleting token", secretInfo.namespace, secretInfo.name)
+			if retriable, err := e.deleteToken(secretInfo.namespace, secretInfo.name, secretInfo.uid); err != nil {
+				glog.Errorf("error deleting serviceaccount token %s/%s for service account %s: %v", secretInfo.namespace, secretInfo.name, secretInfo.saName, err)
+				retry = retriable
+			}
+		default:
+			// Update token if needed
+			if retriable, err := e.generateTokenIfNeeded(sa, secret); err != nil {
+				glog.Errorf("error populating serviceaccount token %s/%s for service account %s: %v", secretInfo.namespace, secretInfo.name, secretInfo.saName, err)
+				retry = retriable
+			}
+		}
 	}
 }
 
-// createSecretIfNeeded makes sure at least one ServiceAccountToken secret exists, and is included in the serviceAccount's Secrets list
-func (e *TokensController) createSecretIfNeeded(serviceAccount *api.ServiceAccount) error {
-	// If the service account references no secrets, short-circuit and create a new one
-	if len(serviceAccount.Secrets) == 0 {
-		return e.createSecret(serviceAccount)
-	}
-
-	// We shouldn't try to validate secret references until the secrets store is synced
-	if !e.secretsSynced() {
-		return nil
-	}
-
-	// If any existing token secrets are referenced by the service account, return
-	allSecrets, err := e.listTokenSecrets(serviceAccount)
+func (e *TokensController) deleteTokens(serviceAccount *api.ServiceAccount) ( /*retry*/ bool, error) {
```

Comment:
> Seems weird to see this instead of a named return value.

Named return values make code that returns early more difficult to reason about, especially when you depend on the default values for those fields.  After the first few bugs, we started avoiding them.